### PR TITLE
Ethan-Challenge-1C

### PR DESCRIPTION
**Pull request by ethanpocock**
Please review this challenge submission

### CODE REVIEW

File: ethan/challenge 1C/challenge1C.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
+def calculator(equation=0):
+
+    import re
```
Comment (bradenrichardson) [inline ethan/challenge 1C/challenge1C.py:3]: Imports should be at top level of file and outside of functions

File: ethan/challenge 1C/challenge1C.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    if equation == 0:
+        equation = input()
+    else:
+        print(equation)
+    equation = re.split(r"([*/+-])",equation)
+
+    # Check for errors in the equation, and force the user to re-input equation if so
+    # # # # # # # # # # # # # # #
+    init_trigger = 0
+    while init_trigger == 0 or ops_count == 0 or num_count == 0 or ops_count >= num_count or gap_check == 1 or invalid_char == 1:
```
Comment (bradenrichardson) [inline ethan/challenge 1C/challenge1C.py:33]: Why have you got this large code block to try to catch unique failures? Should look at implementing try, except code block.

Let the calculation function fail, detect how it fails, create an action for that particular detection.

File: ethan/challenge 1C/challenge1C.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    # Welcome screen text
+    # # # # # # # # # # # # # # #
+    print("Calculatron 4002.")
+    print("---------------------------------------------------")
+    print("SYNTAX. Add: +, Subtract: -, Multiply: *, Divide: /")
+    print("e.g. x + y / z")
+    print("---------------------------------------------------")
+    print("Please type your equation below:")
+
+    # Allow user to input equation if not passed through function
```
Comment (bradenrichardson) [inline ethan/challenge 1C/challenge1C.py:22]: Standardise on one method of entry into your function.

How would an end user interact with this piece of software? Don't give them multiple options, pick one and make that option easy to follow/interact with.

IE less is more

File: ethan/challenge 1C/challenge1C.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                gap_check = 1
+            if re.search(r"[^\s1-9*/+-]",element):
+                invalid_char = 1
+        if ops_count == 0 or num_count == 0 or ops_count >= num_count or gap_check == 1 or invalid_char == 1:
+            print("ERORR.")
+            print("Please type a valid equation with just numbers and at least one of the following operators: +, -, *, /:")
+            equation = input()
+            equation = re.split(r"([*/+-])",equation)
+
+    # Caculate the result
```
Comment (bradenrichardson) [inline ethan/challenge 1C/challenge1C.py:54]: For code sanitation you should split this out into it's own function.

IE. def calculate(input)

File: ethan/challenge 1C/challenge1C.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+            result = ops[key](number1,number2)
+            equation[op_position] = result
+            del equation[num1_position]
+            del equation[num2_position - 1]
+    
+    # Print the result
+    # # # # # # # # # # # # # # #    
+    print("= {}".format(result))
+
+calculator()
```
Comment (bradenrichardson) [inline ethan/challenge 1C/challenge1C.py:72]: Adding on to comment before, file structure should be like this:

import xyz

def input():
    - prompts for input and stores in a way that calculate() can handle
    - handle input related errors here

def calculate(input):
    - takes the output of input and calculates that
    - handle suitable errors here (0*0, math related errors etc)

def main():
    - calculate(input())

if __name__ == '__main__':
    - main()

Comment (bradenrichardson) [reply ethan/challenge 1C/challenge1C.py:72]: https://stackoverflow.com/questions/419163/what-does-if-name-main-do

**Comment by bradenrichardson** [comment]
Looks good! Few comments on the commit

In summary:

1. Split your program into discrete functions (separate input and calculate functions) and enter through main()
2. Understand how your users will interact with your product
3. Understand what errors can occur, and where (which function do they occur in?)
4. Create Try - Except code blocks to catch exceptions/failures
5. Handle your exceptions (What feedback do you provide if you get a ZeroDivisionError vs a ValueError?)

**Comment by ethanpocock** [comment]
> Looks good! Few comments on the commit
> 
> In summary:
> 
>     1. Split your program into discrete functions (separate input and calculate functions) and enter through main()
> 
>     2. Understand how your users will interact with your product
> 
>     3. Understand what errors can occur, and where (which function do they occur in?)
> 
>     4. Create Try - Except code blocks to catch exceptions/failures
> 
>     5. Handle your exceptions (What feedback do you provide if you get a ZeroDivisionError vs a ValueError?)

Thanks mate! Please see the latest commit in line with your feedback.

**Review by bradenrichardson — APPROVED** [review]

**Pull request merged**